src/backend/distributed/master/master_modify_multiple_shards.c

```diff
@@ -0,0 +1,403 @@
+ * the given shard list and returns the hash table containing the connections.
+ * The resulting hash table maps shardId to ShardConnection struct.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIntervalList)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
+
+	ListCell *shardIntervalCell = NULL;
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+
+		OpenConnectionsToShardPlacements(shardId, shardConnectionHash);
+	}
+
+	return shardConnectionHash;
+}
+
+
+/**
+ * OpenConnectionsToShardPlacements opens connections to all placements of the
+ * shard with the given shardId and populates the shardConnectionHash table
+ * accordingly.
+ */
+static void
+OpenConnectionsToShardPlacements(uint64 shardId, HTAB *shardConnectionHash)
+{
+	bool shardConnectionsFound = false;
+
+	/* get existing connections to the shard placements, if any */
+	ShardConnections *shardConnections = GetShardConnections(shardConnectionHash,
+															 shardId,
+															 &shardConnectionsFound);
+
+	List *shardPlacementList = FinalizedShardPlacementList(shardId);
+	ListCell *shardPlacementCell = NULL;
+	List *connectionList = NIL;
+
+	Assert(!shardConnectionsFound);
+
+	if (shardPlacementList == NIL)
+	{
+		ereport(ERROR, (errmsg("Could not find any shard placements for the shard "
+							   UINT64_FORMAT, shardId)));
+	}
+
+	foreach(shardPlacementCell, shardPlacementList)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+			shardPlacementCell);
+		char *workerName = shardPlacement->nodeName;
+		uint32 workerPort = shardPlacement->nodePort;
+		char *nodeUser = CurrentUserName();
+		PGconn *connection = ConnectToNode(workerName, workerPort, nodeUser);
+		TransactionConnection *transactionConnection = NULL;
+
+		if (connection == NULL)
+		{
+			List *abortConnectionList = ConnectionList(shardConnectionHash);
+			CloseConnections(abortConnectionList);
+
+			ereport(ERROR, (errmsg("Could not establish a connection to all "
+								   "placements.")));
+		}
+
+		transactionConnection = palloc0(sizeof(TransactionConnection));
+
+		transactionConnection->connectionId = shardConnections->shardId;
+		transactionConnection->transactionState = TRANSACTION_STATE_INVALID;
+		transactionConnection->connection = connection;
+
+		connectionList = lappend(connectionList, transactionConnection);
+	}
+
+	shardConnections->connectionList = connectionList;
+}
+
+
+/*
+ * SendQueryToPlacement sends the given query string to all given placement
+ * connections of a shard. The query is sent with a BEGIN before the the actual
+ * query so, CommitRemoteTransactions or AbortRemoteTransactions should be
+ * called after all queries have been sent successfully.
+ */
+static int
+SendQueryToConnections(char *shardQueryString, ShardConnections *shardConnections)
```

Comment:
Changed the name. I was not sure because the function sends the given query to already opened placement connections. For a function named SendQueryToPlacements, I would expect something that takes a query and a shard then sends that query to shard placements by opening the connections.